### ISSUE TITLE
Transition to SOLID/DRY Design - Questions and Concerns

### ISSUE THREAD

**Issue by KZeronimo** [simpleinjector/SimpleInjector#324]
Squarely falling into the question camp ...

I've been digesting the general patterns illuminated on Steven's blog [here](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=91) and [here](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=92).  Great stuff and I really like the SOLID principles that are encouraged by the generic interfaces (really seems every application has a layer of `Action<T>` and `Func<T,T>` which is what I first thought of when I saw the `ICommandHandler` and `IQueryHandler` respectively) am I way off base with that though?

My question is simple ... how would logging be added to each decorator?  Steven get's down to a pretty realistic example [here](http://stackoverflow.com/questions/9892137/windsor-pulling-transient-objects-from-the-container/9915056#9915056) ... the measurement and logging is the outermost decorator which makes sense but what if logging was desirable in multiple decorators?  Is is reasonable to maintain a decorator with and without logging - that doesn't seem very DRY.

I'm still early on in grokking this so I appreciate time spent illuminating  these concepts ... logging seems to be cross cutting to all cross cuts!

**Comment by dotnetjunkie** [comment]
> a layer of `Action<T>` and `Func<T,T>` which is what I first thought of when I saw the ICommandHandler and IQueryHandler respectively) am I way off base with that though?

Generic interfaces with one method can be replaced with their `Action<T>` and `Func<T, T>` counterparts, so yes, and this is obviously a quite functional approach (that's the effect of SOLID). The downside of using delegates instead of interfaces is that you are losing the intent of the abstraction (or at least, you do in C#). For instance, does the injected `Action<ShipOrder>` execute the ShipOrder command (as `ICommandHandler<T>` does)? Or does it validate the ShipOrder command (as `IValidator<T>` does)? Or does it run security checks on ShipOrder (as `ISecurityValidator<T>` does)? 

The metadata information that interfaces give makes it easier for us to apply cross-cutting concerns using tooling (Simple Injector for instance) and communicates intend more clearly. Another thing is that C# is an object oriented language and although it contains support for delegates, some things get very inconvenient when working with delegates directly. For instance, while we can easily define an open generic type (e.g. `typeof(LoggingCommandHandlerDecorator<>)`), this is less convenient when we want to point at a generic method. This makes is much harder to construct object graphs in C# if we were using a functional approach using delegates.

Long story short, `ICommandHandler<T>` and `IQueryHandler<T, R>` are the `Action<T>` and `Func<T, R>` equivalence, but with added metadata.

**Comment by dotnetjunkie** [comment]
> but what if logging was desirable in multiple decorators? Is is reasonable to maintain a decorator with and without logging - that doesn't seem very DRY.

Can you show an example of what it is you would like to achieve? Why do you want to have logging in your `TransactionCommandHandlerDecorator<T>`, `AuthorizationCheckerCommandHandlerDecorator<T>` or `CachingQueryHandlerDecorator<TQuery, TResult>`?

**Comment by KZeronimo** [comment]
Certainly like thinking about `ICommandHandler<T>` and `IQueryHandler<T,R>` as the base building blocks for executing code that changes state and retrieves state respectively with the added benefit of metadata ... totally get the perspective about C# being object oriented first - when in Rome ...

So when you compose your solution using these basic building blocks it appears that we will end up with a 1:1 mapping of application actions / use cases to either an `ICommandHandler<T>` or `IQueryHandler<T,R>` instead of a more complex abstraction

For example, when I was introduced to OO it was essentially encapsulation of data and behavior from which I could create inheritance chains for specialization (now we realize that's not the best way to construct software) nonetheless the law of primacy kicks in and I think initially of Customer this way

```
public class Customer
{
	public string Name { get; }
	public string Address { get; }

	public Customer(string name, string address)
	{
		Name = name;
		Address = address;
	}

	public void SaveCustomer()
	{
		// persist the customer info
	}
	
	public void MoveCustomer()
	{
		// update the customer address
	}
	
	public void RetrieveCustomer()
	{
		// retrieve the customer info from persistence
	}
}
```

Now I'm totally on board with the benefits of separating data from behavior and having a `CustomerCommandDTO`  but when it comes to the behaviors now we've lost the organization of the behaviors by having individual interfaces for each Customer operation `ISaveCustomerCommandHandler<T>`, `IMoveCustomerCommandHandler<T>`, `IRetrieveCustomerQueryHandler<TQuery, TResult>` 

I really like the SRP and OCP principles that are all but mandated by this approach - the part that is missing is the higher order service abstraction.  How do we package up a nice Customer Service?  Is it as simple as aggregating the individual handlers into a higher order abstraction?  Is that type of service abstraction still valuable?

I realize I've drifted from the original question and I'll leave it to your discretion on how to proceed; although, I find myself hitting a nice groove with Simple Injector when applying the generic handler pattern with decorators for extension but I have yet to attempt to try and pull things back together into a service which is what this question has ended up being!

Thanks for your time and thoughts

Finally for the later question I certainly don't need logging in all the decorators - and for those think I do perhaps there's a better approach - let me work thru my example and come back with a better formulated question.

**Comment by dotnetjunkie** [comment]
At this point I'm unsure what it is you want to know :-/

**Comment by KZeronimo** [comment]
That's fair ... sorry for formulating the question on the fly

Simply put once you have all the command handlers and query handlers in place that deal with a specific application concern - in this case Customer so for example `ISaveCustomerCommandHandler<T>`, `IMoveCustomerCommandHandler<T>`, `IRetrieveCustomerQueryHandler<TQuery, TResult>` and you wanted to expose those to the developer working on a different component or tier would you simply point them to all the individual handlers (I realize there would be infrastructure to get each handler correctly decorated) that deal with Customer or would you attempt to provide a higher order abstraction around them eg a Customer service?

No worries if my question is still unclear - simply close with a question needs to be specific and better formulated and I'll do some work on my end

**Comment by dotnetjunkie** [comment]
> for example ISaveCustomerCommandHandler<T>, IMoveCustomerCommandHandler<T>, IRetrieveCustomerQueryHandler<TQuery, TResult>

Ah, now I see. This example shows where you go wrong. You should *not* define those specific interfaces; those interfaces are exactly what you should prevent. Instead of having a specific `IShipOrderCommandHandler` the `ICommandHandler<ShipOrderCommand>` is all you need. Consumers can depend on this abstraction and the `ShipOrderCommandHandler` implementation should simply implement `ICommandHandler<ShipOrderCommand>`. When you start using use-case specific interfaces like `IShipOrderCommandHandler` or `IShipOrderCommandHandler<T>` you disallow decorators to be applied and because of this make it much harder to apply AOP. And such specific interface doesn't make it any clearer, since `ICommandHandler<ShipOrderCommand>` is as clear as it gets: it's the handler that will execute the 'ship order' use case.

This might take some time for your developers to get used to, since they are often used to defining many huge interfaces. My experience is however that it takes most developers just a few weeks to get used to these concepts.

So for the majority of components in your system, you just need a handful of interfaces. Here are some examples of interfaces you might define:
 - `ICommandHandler<TCommand>` to process use cases (one-to-one mapping)
 - `IQueryHandler<TQuery, TResult>` to execute queries (one-to-one mapping)
 - `IEventHandler<TEvent>` to execute events that got raised by command handlers and execute them asynchonously (where events are possibly persisted by a queue) (one-to-many mapping)
 - `IValidator<T>` to execute (user) input validations (one-to-many mapping)
 - `IBusinessRule<TEntity>` to execute business rules (one-to-many mapping)
 - `ISecurityValidator<T>` to apply security checks based (one-to-many mapping)

In a system I helped develope a few years ago, we had about 8 or 9 generic abstractions and 2 or 3 non-generic onces which covered about 95% of all components we defined in the system. The advantages are:
 - It communicates really clearly what kind of architectural artifact an object implements.
 - It forces you into a SOLID way of working.
 - It makes it really easy to apply AOP using decorators.
 - It makes it very easy to apply cross-cutting concerns conditionally using generic type constraints.
 - It makes batch registration very easy.
 - It makes your system in to a message passing system and messages can be serialized, deserialized, queued and logged very easily.
 - Makes it trivial to change your two-tier application to a three-tier where commands and query messages are sent over the wire using WCF or Web API.

**Comment by KZeronimo** [comment]
Makes total sense ... I see my mistake ... should be generic interfaces and use case specific implementations!

So armed with generic interfaces `ICommandHandler<TCommand>`, and `IQueryHandler<TQuery, TResult>` with implementations for `MoveCustomerCommandHandler`, `SaveCustomerCommandHandler`, and `GetCustomerByIdQueryHandler`.  As you point out no loss of intent or information with `ICommandHandler<MoveCustomerCommandHandler>` etc ...

I know we could do something like the following ... would you see any value in doing so?

```
public class Customer : ICustomerService
{
	private ICommandHandler<MoveCustomerCommandHandler> _moveCustomerCommandHandler;
	private IQueryHandler<GetCustomerByIdQueryHandler, CustomerDto> _getCustomerByIdQueryHandler;
	.
	.
	.

	public Customer(ICommandHandler<GetCustomerByIdQueryHandler> moveCustomerCommandHandler, IQueryHandler<GetCustomerByIdQueryHandler, CustomerDto> getCustomerByIdQueryHandler)
	{
		_moveCustomerCommandHandler = moveCustomerCommandHandler
		_getCustomerByIdQueryHandler = getCustomerByIdQueryHandler;
        .
		.
		.
    }

	public void MoveCustomer(CustomerDto customer)
	{
		_moveCustomerCommandHandler.Handle(customer);
	}
	
	
	public CustomerDto GetCustomerById(int customerId)
	{
		var query = new CustomerDto { CustomerId = customerId };
		return _getCustomerByIdQueryHandler.Handle(query);
	}
	
	.
	.
	.
}
```
Thanks again ... I'm sufficiently clear and feel confident in closing after your reply ... I've taken enough of your time on this thread

**Comment by dotnetjunkie** [comment]
You absolutely can create an `ICustomerService` with `CustomerServiceImpl` implementation, but that doesn't do anything but add an extra layer of indirection. So instead of injecting `ICustomerService` into consumers (like controllers and other parts of the system), inject `ICommandHandler<T>` abstractions directly into the consumer. You might be tempted to create these 'aggregate abstractions' when a consumer requires a dozen of `IQueryHandler` and `ICommandHandler` dependencies, but in that case your consumer *is* violating the Single Responsibility Principle; it is doing way too much.

Although wrapping handlers in an `ICustomerService` 'aggregate abstraction' only adds to the number of layers of indirection, the other way around could work well as an intermediate step. For instance, I sometimes have to deal with an legacy code base that contains a ton of [ISP voilating](https://en.wikipedia.org/wiki/Interface_segregation_principle) `IXXXService` interfaces. Refactoring this all at once is usually a bad idea, but still we need to move forward and need a more maintainable way to apply cross cutting concerns. What I might do is let consumers depend on an `ICommandHandler<MakeCustomerPrefered>` abstraction and have its `MakeCustomerPreferedHandler` implementation depend on the old `ICustomerService`. This way we can hide the old `ICustomerService` interface from the consumer and directly start applying cross-cutting concerns around the `ICommandHandler<T>` without having to refactor the `CustomerServiceImpl`. Eventually I want to get rid of these big SRP, OCP and ISP violating `XXXService` implementations, but we can do this one step at the time, while new functionality can be developed directly using the SOLID handler approach.

Do note that refactoring an existing code base is *always* a lot of work, no matter which new design you select. Often cross-cutting concerns are so interweaved with the legacy code that refactoring is hard, and it confuses developers, because they will project the old solution onto the new model. While I find most developers picking up the whole concept rapidly when we're building an application from scratch, this isn't always the case for developers who try to apply these concepts to an existing code base.

**Comment by KZeronimo** [comment]
Thanks for the in depth reply

> Do note that refactoring an existing code base is always a lot of work, no matter which new design you select. Often cross-cutting concerns are so interweaved with the legacy code that refactoring is hard, and it confuses developers, because they will project the old solution onto the new model. While I find most developers picking up the whole concept rapidly when we're building an application from scratch, this isn't always the case for developers who try to apply these concepts to an existing code base.

Absolutely!  You really helped me crystallize the concerns I have and how to help folks thru [Trough of Disillusionment](https://en.wikipedia.org/wiki/Hype_cycle) in making this transition to SOLID/DRY design (as you can see I've changed the name of the issue to better describe our dialog).

1. We arrive at design with the generic `ICommandHandler<T>` and `IQueryHandler<T, R>` interfaces and SOLID implementations of `XCommands` and `XQueries` and we can even stay DRY with our decorators for cross-cutting concerns.  This should be an easy sell and transition on the maintainability and testing axis; however, the concern is for the consumer and taking away their 'nice ' `IXYZService` interface and so you saw my attempt to meet that need with the 'aggregate abstraction'.  This is just something that needs to be acknowledged and worked thru with each team and the dynamic therein - I'm resolved and fine with that

> Often cross-cutting concerns are so interweaved with the legacy code that refactoring is hard ...

2. Yes!  Now I have my logging question better formulated ... this cross-cutting concern is everywhere and I can easily concede we log too much.  Again, it's how to help convince that moving monitoring and logging to decorators is equivalent to the liberal use of `Stopwatch` and `Log` in a SRP violating codebase .  The point is it won't be equivalent and it feels like a concession.  The compromise is that if a particular responsibility really needs monitoring and logging  go ahead an implement as such.

So no real question this time Steven just processing the insights you've provided

Thank you for the dialog ... I've benefited immensely
